### ISSUE TITLE
Syntax error in PL/SQL block when the last token of an SQL statement is `end`

### ISSUE THREAD

**Issue by PhilippSalvisberg** [IslandSQL/IslandSQL#203]
Here's an example:

```sql
declare
   l_count integer;
begin
   select count(*)
     into l_count
     from user_tables end;
end;
/
```

Syntax error is `extraneous input '<EOF>' expecting ... on Line 8 Col 2`.

The next image shows what is wrongly hidden by the scope lexer:

![image](https://github.com/user-attachments/assets/2878e811-3d21-4b43-9366-c564f52b005e)

The `end` in line 6 is interpreted as the end of the anonymous PL/SQL block. As a result, the rest of the file is not passed to the parser, leading to the syntax error.

**Comment by PhilippSalvisberg** [comment]
You might think that we could fix this in the scope lexer by adding the following after [line 601](https://github.com/IslandSQL/IslandSQL/blob/v0.13.0/src/main/antlr4/ch/islandsql/grammar/IslandSqlScopeLexer.g4#L601) :

```
// handle everything that ends on a semicolon with or without an end keyword before the semicolon
CB_SELECT_INTO: 'into' -> more, pushMode(TO_SEMI_MODE);
```

and adding a `TO_SEMI_MODE` like this:

```
mode TO_SEMI_MODE;

// fail-safe, process tokens that are waiting to be assigned after "more"
SEMI_EOF: EOF -> popMode;

SEMI_SEMI: ';'  -> popMode;

SEMI_SELECTION_DIRECTIVE_START: '$if' -> more, pushMode(CONDITIONAL_COMPILATION_MODE);

SEMI_ML_COMMENT: (ML_COMMENT|ML_COMMENT_ORCL) -> more;
SEMI_SL_COMMENT: SL_COMMENT -> more;
SEMI_WS: WS -> more;
SEMI_STRING: STRING -> more;
SEMI_ID: ID -> more;
SEMI_QUOTED_ID: QUOTED_ID -> more;
SEMI_ANY_OTHER: . -> more;
```

However, this approach leads to many other problems. For example, with [iterators](https://docs.oracle.com/en/database/oracle/oracle-database/23/lnpls/iterator.html), using merge or insert statements (which also use the `into` keyword).

In the end, this is a limitation of the current implementation of the current [hideOutOfScopeTokens](https://github.com/IslandSQL/IslandSQL/blob/v0.13.0/src/main/java/ch/islandsql/grammar/util/TokenStreamUtil.java#L46) implementation which is based on an ANTLR4 lexer. I am not sure if it is possible to lift this restriction within the lexer with reasonable effort.

We will therefore document this behaviour as a limitation.

**Comment by PhilippSalvisberg** [comment]
closed via #204
